index.d.ts

```diff
@@ -16,16 +16,80 @@
 import {Node} from 'unist'
 import {VFile, VFileCompatible} from 'vfile'
 
+/* eslint-disable @typescript-eslint/ban-types */
+
+type VFileWithOutput<Result> = Result extends string
+  ? VFile
+  : Result extends Uint8Array
+  ? VFile
+  : Result extends object
+  ? VFile & {result: Result}
+  : VFile
+
+type UsePlugin<
+  ParseTree extends Node,
+  CurrentTree extends Node,
+  CompileTree extends Node,
+  CompileResult,
+  Input,
+  Output
+> = Output extends Node
+  ? Input extends string
+    ? // If `Input` is `string` and `Output` is `Node`, then this plugin
+      // defines a parser, so set `ParseTree`.
+      // This also assumes that this is the first plugin (so no current tree is
+      // or compile tree is defined yet), so set those too.
+      Processor<Output, Output, Output, CompileResult>
+    : Input extends Node
+    ? // If `Input` is `Node` and `Output` is `Node`, then this plugin defines a
+      // transformer, its output defines the input of the next, so set
+      // `CurrentTree`.
+      Processor<ParseTree, Output, CompileTree, CompileResult>
+    : // Else, `Input` is something else and `Output` is `Node`:
+      never
+  : Input extends Node
+  ? // If `Input` is `Node` and `Output` is not a `Node`, then this plugin
+    // defines a compiler, so set `CompileTree` and `CompileResult`
+    Processor<ParseTree, CurrentTree, Input, Output>
+  : // Else, `Input` is not a `Node` and `Output` is not a `Node`.
+    // Maybe it’s untyped, or the plugin throws an error (`never`), so lets
+    // just keep it as it was.
+    Processor<ParseTree, CurrentTree, CompileTree, CompileResult>
```

Comment:
interesting use of inference to extract generics.
Is this approach used to overcome quirks in JSDoc generics?
Are there other benefits?

Reply:
How else could it be done?

Reply:
`Processor<Input>` each `Plugin<Input>` must match the Processor, `use()` creates a new `Processor<Input>` with the output of the current step in the chain feeding forward.

Reply:
that’s done in:

```ts
export type Plugin<
  PluginParameters extends any[] = any[],
  Input = Node,
  Output = Input
> = (
```

And the `use` overloads:

```ts
use<
  PluginParameters extends any[] = any[],
  Input = Specific<Node, CurrentTree>,
  Output = Input
>
```

…but this conditional makes a *processor*, based on how one is currently configured, and applying those `Input` / `Output` from a plugin